shell/src/Luna/Shell.hs

```diff
@@ -139,33 +147,129 @@ stdlibPath = do
+                   $ view importedClasses imp
+    in flip map (Map.elems functions ++ methods) $ \case
+        Left errors -> Just errors
+        Right     _ -> Nothing
+
+forceCompilation :: Project.CompiledModules -> IO ()
+forceCompilation (Project.CompiledModules m p) = do
+    let !errors = P.concat $ map forceImports (p : Map.elems m)
+    case catMaybes errors of
+        []  -> return ()
+        err -> throwM $ CompilationError $ P.concat err
+
+data CompilationError = CompilationError [CompileError]
+    deriving Show
+
+instance Exception CompilationError where
+    displayException (CompilationError errors) =
+        convert $ Layout.concatLineBlock $ Layout.render $ formatErrors errors
+
+data Options = Options { _projectDirectory      :: Maybe FilePath
+                       , _exhaustiveCompilation :: Bool
+                       , _fileWatch             :: Bool
+                       }
+
+makeLenses ''Options
+
+options :: OptParse.Parser Options
+options = Options
+      <$> OptParse.optional (OptParse.argument OptParse.str (
+              OptParse.metavar "PROJECT_DIR"
+           <> OptParse.help "Path to a project"))
+      <*> OptParse.switch (
+              OptParse.long "exhaustive"
+           <> OptParse.help ("Enables exhaustive compilation "
+                          <> "of all sources and functions"))
+      <*> OptParse.switch (
+              OptParse.long "file-watch"
+           <> OptParse.help "Enables recompilation on a file change")
+
+
 main :: IO ()
-main = do
-    mainPath'    <- getCurrentDirectory
-    mainPath     <- Path.parseAbsDir mainPath'
-    let mainName  = Project.getProjectName mainPath
-    stdPath      <- stdlibPath
-    (_, std)     <- Project.prepareStdlib  (Map.fromList [("Std", stdPath)])
+main = shell =<< OptParse.execParser opts
+    where
+        opts = OptParse.info (options <**> OptParse.helper)
+            (OptParse.fullDesc <> OptParse.header "luna compiler")
+
+
+evaluateMain :: Maybe (Either [CompileError] Function.Function) -> IO ()
+evaluateMain mainFun = case mainFun of
+    Just (Left e)  -> do
+        putStrLn "Luna encountered the following compilation errors:"
+        Terminal.putStrLn $ Layout.concatLineBlock $ Layout.render $
+            formatErrors e
+        putStrLn ""
+        liftIO $ putStrLn "Compilation failed."
+    Just (Right f) -> do
+        putStrLn "Running main..."
+        res <- liftIO $ runIO $ runError $ LunaValue.force $
+            f ^. Function.value
+        case res of
+            Left err -> putStrLn $ "Luna encountered runtime error: " ++ err
+            _        -> putStrLn "main finished"
+    Nothing -> putStrLn "Function main not found in module Main."
+
+shell :: Options -> IO ()
+shell opts = do
+    mainPath' <- maybe getCurrentDirectory return $ opts ^. projectDirectory
+    mainPath  <- Path.parseAbsDir mainPath'
+    let mainName = Project.getProjectName mainPath
+    stdPath   <- stdlibPath
```

Comment:
Purely a nitpick, but if `mainPath'` --> `mainPath`, then `stdPath'` --> `stdPath` :D Also, I think `Path.parseAbsDir =<< stdlibPath` would be even more readable (one man's opinion).

Reply:
I agree that it's annoying, but both are required in different places and combining respective variables into one would lengthen some of the latter lines, causing them to be split.

Reply:
Maybe we should give them better names? By looking at the code fast I dont know what they are for.